### DISCUSSION TITLE
도메인간 의존성에 관하여

### DISCUSSION THREAD

**Discussion by lcqff** [BDD-CLUB/01-doo-re-back#185] [논의 사항]
https://github.com/BDD-CLUB/01-doo-re-back/pull/184#discussion_r1664081355

미나님이 남겨주신 해당 리뷰를 보고 Doore 코드의 의존관계에 대한 생각을 해보았습니다! 

특정 도메인의 서비스레이어가 다른 도메인의 서비스코드를 호출한다는 점에 초점을 맞춰 도메인간 분리가 가능할지 의문이라 답했는데, 오늘 다시 한 번 생각해보니 서비스레이어간 분리가 아니라 **레파지토리 레이어 관점에서의 도메인 분리**가 필요할거 같다는 생각이 들었습니다.

예를 들자면 아래와 같은 코드가 있습니다.

```java
   public void createTeam(final TeamCreateRequest request, final MultipartFile file, final Long memberId) {
        Member member = validateExistMember(memberId);
        final Team team = Team.builder()
                .name(request.name())
                .description(request.description())
                .imageUrl(imageUrl)
                .build();
        teamRepository.save(team);

        final TeamRole teamRole = TeamRole.builder()
                .memberId(memberId)
                .teamId(team.getId())
                .teamRoleType(ROLE_팀장)
                .build();
        teamRoleRepository.save(teamRole);

        final MemberTeam memberTeam = MemberTeam.builder()
                .member(member)
                .isDeleted(false)
                .teamId(team.getId())
                .build();
        memberTeamRepository.save(memberTeam);
    }
``` 
(이해의 편의를 위해 FILE과 관련된 코드는 삭제했습니다.)

위 코드는 **Team 도메인의 서비스 코드**입니다.
그러나 Team 뿐만 아니라 다른 도메인의 Member와 memberTeam, TeamRole을 모두 날것의 엔티티 그대로 조회하고 있습니다.
이는 Team 도메인에서 더티체킹으로 인해 **다른 도메인의 엔티티의 상태를 변경을 발생시키는 실수**가 일어날 수 있다는 뜻입니다.
실제로 Team 도메인에서 TeamRole, memberTeam 도메인의 상태를 변경시키고 있고요

저희는 한 **도메인이 자신의 역할에만 집중**할 수 있도록 **간접 참조**를 사용하고 있습니다.

```java
public class Team extends BaseEntity {

    private static final int MAX_LENGTH = 250;

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, length = MAX_LENGTH)
    private String name;

    @Column(nullable = false, length = MAX_LENGTH)
    private String description;

    @Column(nullable = false, length = MAX_LENGTH)
    private String imageUrl;

    @Column(nullable = false)
    private Boolean isDeleted;
}
``` 
(Team은 MemberTeam과 다대다 관계에 있습니다.)

그러나 위와 같이 Team 도메인이 자신과 관계없는 Member, memberTeam, TeamRole의 다른 도메인의 엔티티를 조회하여 실수의 가능성을 높이고, 실제 변경을 일으키는 것은 설계 의도와 맞지 않다 생각합니다.

제가 도메인간 의존성을 낮추고, 신뢰성있는 엔티티를 만들기 위해 제안하는 방법은 아래와 같습니다.

```java
   public void createTeam(final TeamCreateRequest request, final MultipartFile file, final Long memberId) {
        MemberResponseDto memberDto = MemberQueryService.validateExistMember(memberId);
        
       final Team team = Team.builder()
                .name(request.name())
                .description(request.description())
                .imageUrl(imageUrl)
                .build();
        teamRepository.save(team);

        teamRoleCommandService.assignTeamLeaderRole(memberId, teamId);

       memberTeamCommandService.addMemerTeam(memberId, teamId);
    }
``` 

이러한 방식을 사용하면 도메인간 독립성을 높힐 수 있을 뿐더러 캡슐화를 시킬수도 있고, 엔티티가 신뢰성을 가지며, validateExistMember와 같이 여러 도메인에서 사용하고 있는 코드이 중복을 줄일수 있다는 장점이 있습니다.

다른분들의 의견이 궁금합니다! 

---

Doore에 대한 설계를 시작할때는 제가 백엔드에 막 입문한 뉴비였던 터라 잘 몰랐는데 그때 저희가 사용한 `도메인형 패키지 구조`와 설계방식이 DDD 모델과 굉장히 흡사한거 같더라구요

 DDD 모델을 적용해 애그리거트를 사용하는것도 도메인 간 분리에 크게 도움이 될 거 같은데 이 부분은 구조가 전체적으로 수정될거 같아, 일단 서로다른 도메인의 서비스레이어와 레파지토리가 분리가 된 이후 의논해보는 것도 좋을 거 같습니다. 
https://mangkyu.tistory.com/318

**Comment by JJimini** [comment]
글 잘 읽었습니다!😄 여유로울 때 글을 자세히 읽어보고 싶어 comment가 늦었습니다..!
제가 도메인의 의존성에 대해 고민했던 부분은 **서비스 간 메서드를 호출할 시 강한 결합**이 생기는 부분입니다. 소프트웨어 설계의 기본이자 중요한 목표인 "**낮은 결합도, 높은 응집도**" 부분에 어긋나는 부분이라고 생각합니다.

팜님의 의견처럼 해당 방식은 **코드의 중복을 줄이고 중앙 집중형 방식이 되는 장점**이 있으나, **강한 결합이 생기고 복잡성**이 증가합니다. 모종의 이유로 메서드에 문제가 생긴다면 그 메서드를 사용하는 모든 메서드가 동작하지 않습니다..

다음으로 본문의 방식은 **중복을 피하고자 했으나 또 다른 중복**을 불러일으킵니다. 예를 들어 `saveParticipant`를 호출하면 이미 진행한 본인 검증, 팀 검증 등 검증이 2번 일어나게 됩니다.(그렇다고 본래 `saveParticipant`코드에 검증 과정을 뺄 수도 없으니..) 또한 메서드를 작성할 때, **메서드의 역할이 뿐만 아니라 이 메서드가 다른 메서드에 의해 호출되는 상황까지 고려**하여 general한 느낌으로 작성해야 하며, 이러한 부분은 메서드 유연하지 않게 작성될 수 있습니다!

또한 신뢰성 있는 엔티티는 이해하였지만, **의존성을 낮추었다는 말에는 모순이 존재**한다고 생각합니다.
각 도메인 안에서 의존하고 그것을 다른 도메인의 호출로 엮는다는 것은 컴파일 타임 의존성이 높아지고 이것은 결합도가 올라가는 방법 입니다.

추가로 캡슐화가 된다고 하셨는데, 클래스 외부에서 접근하려면 모든 메서드를 public으로 풀어야 합니다. 그렇다면 **데이터 은닉이 제대로 되지 않는 문제**가 생길 것 같습니다. 그리고 **순환 참조** 문제가 발생할 수 있는 위험이 있습니다.

정리하면 해당 방법은 도메인 간 의존성을 낮추고자 하였으나 컴파일 타임 의존성이 증가하는 방법인 것 같습니다.
저는 강한 결합도가 생기는 프로젝트는 좋지 않은 프로젝트라고 생각합니다. 비록 저희가 권한을 서비스 레이어에서 처리하고, 간접참조를 사용함으로써 불필요하게 중복되는 코드와 유지보수 시 어려움이 생기긴했지만... Keeper 프로젝트를 비롯하여 타 프로젝트 코드를 보면 서비스 간 호출을 하지 않는 것도 (S3와 같은 특이 케이스 제외 -> 보통 이런 경우 global 도메인을 만들어서 구분함)참고할 수 있을 것 같습니다!
(저희 같은 경우 crop도 특이 케이스가 될 수 있겠네요..!)

**Comment by JJimini** [comment]
추가로 코드가 중복되는 부분은 모듈 서비스 클래스 개발을 통해 해결할 수 있을 것 같습니다!
현재 두레는 검증 관련 중복 코드가 많습니다. 모듈 서비스를 해당 도메인의 DAO만 의존하게 하여 개발하면 위의 문제를 어느 정도 해결하면서 중복 문제를 효과적으로 해결할 수 있을 것 같습니다. commandService, QueryService를 호출하는 것이 아닌 해당 역할을 하는 class를 따로 만들어 호출하는 것이죠

- Module Service는 해당 도메인의 DAO만을 의존
- 예를 들어 Module Service에 들어갈 내용은 검증과 같은 부분이 들어감 (기능 X)
- Module Service는 convenience folder로 구분

예를 들어 해당 코드를 리팩토링하면 다음과 같습니다.
- `member `- `application `- `convenience `- `MemberFindService`
``` java
    public Member findById(final Long memberId){
        return memberRepository.findById(memberId).orElseThrow(()-> new MemberException(NOT_FOUND_MEMBER));
}
```

- `team `- `application `- `TeamCommandService`
``` java
   public void createTeam(final TeamCreateRequest request, final MultipartFile file, final Long memberId) {
        Member member = memberFindService.findById(memberId);
        final Team team = Team.builder()
                .name(request.name())
                .description(request.description())
                .imageUrl(imageUrl)
                .build();
        teamRepository.save(team);
        createMemberTeam(member, team.getId());
        assignTeamRole(team.getId(), memberId);
    }
    (createMemberTeam, assignTeamRole코드는 생략)

**Comment by lcqff** [comment]
답변을 요약했을때 제가 제시한 방법을 사용할시 아래와 같은 문제가 발생할 수 있다고 이해했습니다.

1. 서로 다른 서비스간 메서드 호출시 강한 결합과 복잡성이 발생한다.
2. 특정 서비스의 메서드에 문제가 발생한다면 해당 메서드를 사용하는 다른 서비스에도 문제가 발생한다. 
3. 여전히 중복이 발생한다.
4. 해당 메서드가 다른 메서드에 의해 호출될 수 있음을 고려하며 코드를 작성하면 메서드의 유연성을 잃는다.
5. 다른 서비스의 메서드를 호출하는 것은 의존성을 낮추지 않는다. (이는 1번과 함께 설명하겠습니다.)
6. 다른 서비스 메서드를 public으로 함으로 인해 데이터 은닉이 되지 않는 문제가 발생한다.
7. 서비스 레이어간 참조로 발생하는 순환 참조 문제가 발생할 수 있다.

또한 미나님의 의견은 Service에서 다른 도메인의 Service의 참조를 지양하고 대신 Repository를 참조하자는 것으로 이해했습니다.

## 의존성 & 결합도에 대해

- 서로 다른 서비스간 메서드 호출시 강한 결합과 복잡성이 발생한다.
- 다른 서비스의 메서드를 호출하는 것은 의존성을 낮추지 않는다.

위 두 의견은 서료 연관지어 설명할 수 있을 거 같아 함께 설명하겠습니다.

### 의존성이란

*자바 객체 지향 원리*에서는 의존성에 대해 아래와 같이 설명합니다.

> ***의사코드**
운전자가 자동차를 생산한다. 자동차는 내부적으로 타이어를 생산한다.*
> 

> ***자바 표현**
new Car();
Car 객체생성자에서 newTire():*
> 

> ***그리고 의존성을 단순하게 정의하면 다음과 같다.***
> 
> 
> *의존성은 new 다.* 
> 
> *new를 실행하는 Car와 Tire 사이에서 Car가 Tire에 의존한다.*
> 

*…(본문)…*

*DI를 마무리하기 전에 마지막으로 언급할 사항이 하나있다. 의존관계가 new라고 단순화했던 부분이다.*

*사실 변수에 값을 할당하는 모든 곳에 의존관계가 생긴다. 즉, **대입연산자(=)에 의해 변수에 값이 할당되는 순간에 의존이 생긴다**. 변수가 지역 변수이 건 속성이건, 할당되는 값이 리터럴이건 객체이건 의존은 발생한다.* 

*의존 대상이 내부에 있을 수도 있고, 외부에 있을 수도 있다.* 

*DI는 외부에 있는 의존대상을 주입하는 것을 말한다. 의존대상을 구현하고 배치할 때 SOLID와 응집도는 높이고 결합도는 낮추라는 기본원칙에 충실해야 한다. 그래야 프로젝트의 구현과 유지보수가 수월해진다.*

---

또한 *<개발자가 반드시 정복해야 할 객체 지향과 디자인 패턴>*에서는 의존성을 아래와 같이 설명합니다.

```java
public class FlowController {
	public void process() {
		FileDataReader reader = new FileDataReader(fileName); //객체 생성
		byte[] plainBytes = reader.read(); // 메서드 호출
		
		...
	}
}

```

*이렇게(*위 코드와 같이) 한 객체가 다른 객체를 생성하거나 다른 객체의 메서드를 호출할 때 이를 그 객체에 의존(dependency)한다고 한다. 
….(생략)*

*객체를 생성하든 메서드를 호출하든 또는 파라미터로 전달받든 다른 타입에 의존을 한다는 것은 의존하는 타입에 변경이 발생할 때 나도 함께 변경될 가능성이 높다는 것을 뜻한다.*  

---

아래는 제가 디스커션에서 제시했던 수정 전 코드 & 수정 후 코드입니다.

```java
   public void createTeam(final TeamCreateRequest request, final MultipartFile file, final Long memberId) {
        Member member = validateExistMember(memberId);
        final Team team = Team.builder()
                .name(request.name())
                .description(request.description())
                .imageUrl(imageUrl)
                .build();
        teamRepository.save(team);

        final TeamRole teamRole = TeamRole.builder()
                .memberId(memberId)
                .teamId(team.getId())
                .teamRoleType(ROLE_팀장)
                .build();
        teamRoleRepository.save(teamRole);

        final MemberTeam memberTeam = MemberTeam.builder()
                .member(member)
                .isDeleted(false)
                .teamId(team.getId())
                .build();
        memberTeamRepository.save(memberTeam);
    }
```
(수정 전)

```java
   public void createTeam(final TeamCreateRequest request, final MultipartFile file, final Long memberId) {
        MemberResponseDto memberDto = MemberQueryService.validateExistMember(memberId);
        
       final Team team = Team.builder()
                .name(request.name())
                .description(request.description())
                .imageUrl(imageUrl)
                .build();
        teamRepository.save(team);

        teamRoleCommandService.assignTeamLeaderRole(memberId, teamId);

       memberTeamCommandService.addMemerTeam(memberId, teamId);
    }
```
(수정 후)

우선 제가 언급한 의존성이 낮아진다라는 말은 다른 도메인에서의 불필요한 객체 할당이 사라지며 다른 도메인의 객체에 대한 의존성이 줄어든다는 뜻이었습니다. 

 

### 서비스 레이어를 참조하며 발생하는 의존성

그러나 미나님의 의견을 읽어봤을 때 객체보다는 **서비스 레이어를 참조하는 것에서 발생한 의존성**을 말씀하시는 것 같아 이에 대해서도 설명하겠습니다.

발췌한 위 두 책 중 _<자바 객체 지향 원리>_에서는 이러한 의존성으로 발생하는 결합도를 낮추는 방법으로 **의존성 주입(Dependency  Injection)**에 대해, _<개발자가 반드시 정복해야 할 객체 지향과 디자인 패턴>_에서는 의존성으로 발생할 수 있는 구현 변경의 비유연성에 대해 경고하며 이에 대한 해결 방안으로 **캡슐화**에 대해 설명합니다.

짧게 말하자면 Service 레이어의 의존성과 그 의존성으로 인해 발생하는 결합도는 어쩔수 없이 발생하는 것이라는 생각합니다. 미나님이 제시하신 방법도 결국 그 의존 대상이 Service에서 Repository로 바뀌었을 뿐 여전히 의존이 발생하고 있으니까요. 

 

![image](https://github.com/user-attachments/assets/b2b08e27-a38d-403a-8294-9f9e906fa0ff)

계층형 설계를 따르며 코드 구조를 작성한 만큼 하위 계층이 아닌 동일 계층의 Service끼리 의존성을 가지게 되는 것에 대해 거부감을 느낄 수도 있겠다는 생각이 듭니다. 상위 계층은 하위 계층에 의존해야하며 상위 계층을 알지 못해야 한다는 이야기만 들어왔지 동일한 계층끼리의 의존이 적절한가 부적절한가에 대한 이야기는 들어본적 없으니까요. 저도 예전에 이 부분에 대해 의문을 가져 관련된 내용을 찾아본적이 있습니다. 

DAO에 대한 참조는 있어선 안된다!라 주장하는 것은 아니고, 프로젝트의 성향에 따라 Service가 아닌 Repository만을 참조하는 것이 좋은 설계일 수도 있을 거라 생각합니다.(미나님의 다른 프로젝트가 그러하듯이)

Keeper 코드 중 Repository만을 참조한 Service는 달리 규칙이 있어 그렇게 코드를 작성한 것이 아니라 그냥 다른 도메인간 의존할 일이 없었다는 대답을 들었습니다. Keeper 코드에서도 필요에 따라선 다른 도메인의 Service를 참조해 오기도 했고요. 사실 Service에서 다른 Service를 참조해오는 것은 Keeper가 아니라 타 프로젝트에서도 흔히 볼 수 있는 방식입니다.

그러나 Serivce에서 Service를 참조해오는 것이 좋다/Repository를 참조하는 것이 좋다에 관해서는 정답이 없는 부분이라 다른 프로젝트를 예시로 들기보다는 Doore의 상황에 따라 판단하는 것이 좋다 생각합니다.

일단 제 의견은 DooRe처럼 도메인간 확실한 상호작용이 필요한 코드에서는 어떤 도메인에 대한 기능이 해당 도메인에서만 일어나도록 관심사를 분리하는 것이 좋다는 것입니다.

Service 레이어가 다른 Service를 의존해도 되는가에 대한 부분은 사실 제 의견보다는 전문가? 분들의 말씀이 더 신뢰가 갈 거 같아 링크 몇개를 첨부해봅니다,,,,

[서비스 레이어 끼리의 의존 - 인프런](https://www.inflearn.com/questions/987690/%EC%84%9C%EB%B9%84%EC%8A%A4-%EB%A0%88%EC%9D%B4%EC%96%B4-%EB%81%BC%EB%A6%AC%EC%9D%98-%EC%9D%98%EC%A1%B4)

[의존성관련 질문 - 인프런](https://www.inflearn.com/questions/829223/%EC%9D%98%EC%A1%B4%EC%84%B1%EA%B4%80%EB%A0%A8-%EC%A7%88%EB%AC%B8)

(사실 제가 하고 싶은 말은 여기까지입니다. 아래는 미나님이 의문을 가지신 만큼 설명은 해드려야 할거 같아 일단 적어두긴 했지만 크게 중요한 부분은 아닙니다…)

<br>
<br>

## 특정 서비스의 메서드에서 발생한 문제가 해당 메서드를 사용한 다른 서비스에 영향을 미치는 경우에 대해

- 특정 서비스의 메서드에 문제가 발생한다면 해당 메서드를 사용하는 다른 서비스에도 문제가 발생한다.

```java
   public void createTeam(final TeamCreateRequest request, final MultipartFile file, final Long memberId) {
        MemberResponseDto memberDto = MemberQueryService.validateExistMember(memberId);
        
       final Team team = Team.builder()
                .name(request.name())
                .description(request.description())
                .imageUrl(imageUrl)
                .build();
        teamRepository.save(team);

        teamRoleCommandService.assignTeamLeaderRole(memberId, teamId);

       memberTeamCommandService.addMemerTeam(memberId, teamId);
    }
```


이 부분은 `assignTeamLeaderRole`이나 `addMemerTeam`와 같이 다른 서비스에 존재하는 메서드에서 발생한 문제가 `createTeam`에 영향을 끼친다고 이해했습니다.

```java
   public void createTeam(final TeamCreateRequest request, final MultipartFile file, final Long memberId) {
        Member member = memberFindService.findById(memberId);
        final Team team = Team.builder()
                .name(request.name())
                .description(request.description())
                .imageUrl(imageUrl)
                .build();
        teamRepository.save(team);
        createMemberTeam(member, team.getId());
        assignTeamRole(team.getId(), memberId);
    }
    (createMemberTeam, assignTeamRole코드는 생략)
```
(미나님이 제시하신 코드)

미나님이 제시해주신 코드의 `createMemberTeam`와 `assignTeamRole`은 기존 코드에서 해당 역할을 하는 코드를 동일한 서비스에 private한 메서드로 분리한 것이라 이해했습니다. 

미나님이 제시하신 코드를 따른다면 확실히 memberTeamCommandService의 `addMemerTeam`나 teamRoleCommandService의 `assignTeamLeaderRole`에 문제가 발생하더라도 `createTeam`에 문제가 발생하진 않습니다. `createMemberTeam`와 `assignTeamRole`은 TeamCommandService의 코드이니까요.

그러나 이는 동일한 기능에 대한 관리 포인트를 나누게 됩니다. 만일 권한을 지정하는 코드나 팀원을 생성하는 코드의 로직이 수정될 경우 저희는 해당 기능이 존재하는 모든 Service코드를 찾아 직접 로직을 수정해야 할 것입니다.

또한 이러한 관리포인트의 분할은 오히려 문제의 발생 가능성을 높인다고 생각합니다. 미나님의 말씀대로 `createTeam`에서  teamRoleCommandService의 권한 설정 코드를 사용하는 경우 teamRoleCommandService의 권한 설정 메서드에 문제가 발생한다면 `createTeam` 또한 문제가 발생할 것이기에 `createTeam`을 실행시키는 것으로 문제를 포착할 수 있습니다. 문제가 되는 코드가 실제 개발 환경에 배포될 가능성도 현저히 줄어듭니다. 

그러나 동일한 기능을 하는 메서드가 여러 Service에 중복되어 작성되어 있는 경우 한 Service의 정상 작동이 다른 모든 기능이 정상 작동을 보증할 수 없습니다. 권한 기능의 작동을 보장하기 위해서는 해당 기능이 작성되있는 모든 Service를 찾아 코드를 확인해야 할 것입니다.

## 여전한 중복의 발생

```java
    public void createStudy(final StudyCreateRequest request, final Long teamId, final Long memberId) {
        validateExistMember(memberId); //회원 검증
        validateExistTeam(teamId);
        checkEndDateValid(request.startDate(), request.endDate());
        final Study study = studyRepository.save(request.toStudy(teamId));
        studyRoleRepository.save(StudyRole.builder()
                .studyRoleType(ROLE_스터디장)
                .studyId(study.getId())
                .memberId(memberId)
                .build());
        participantCommandService.saveParticipant(study.getId(), memberId, memberId);
    }
```
(StudyCommandService)

```java
    public void saveParticipant(final Long studyId, final Long memberId, final Long studyLeaderId) {
        validateExistStudyLeader(studyId, studyLeaderId);
        validateExistStudy(studyId);
        final Member member = validateExistMember(memberId); //회원 검증이 두 번 발생한다.
        final Participant participant = Participant.builder()
                .studyId(studyId)
                .member(member)
                .build();
        participantRepository.save(participant);
        assignParticipantRole(studyId, memberId, studyLeaderId);
    }
```
(PartcipantCommandService)


이 부분은 제가 이해를 잘 했는지 모르겠습니다. `createStudy`의 회원 검증을 삭제하는 것으로 해결되는 문제 아닌가요? 다른 도메인인 participantCommandService을 사용해서 발생한 문제는 아니라고 생각됩니다.

미나님이 제시한 대로라면 StudyService 아래에 `saveParticipant` 메서드를 생성하는 것일텐데 해당 방식과 현재 방식은 크게 다르지 않다고 생각됩니다. (현재 방식으로는 `saveParticipant` 내용을 확인하기 위해 한단계 들어가야 한다는 번거로움 정도?)

오히려 현재 방식을 유지하며 도메인간 의존성 방향을 정하는 것이 검증을 중복해서 발생시킬 실수를 줄이지 않을까 생각이 듭니다.(Study→Participant로 의존성 방향을 잡아 최종적인 검증은 participant에서 발생하는 것으로)(물론 participant를 사용하지 않는 다른 study 메서드에는 해당되는 말이 아닙니다.)

## 메서드의 유연성에 관해

- 해당 메서드가 다른 메서드에 의해 호출될 수 있음을 고려하며 코드를 작성하면 메서드의 유연성을 잃는다.

_<개발자가 반드시 정복해야 할 객체 지향과 디자인 패턴>_에서는 의존성으로 발생할 수 있는 구현 변경의 비유연성에 대해 경고하며 이에 대한 해결 방안으로 **캡슐화**에 대해 설명합니다.

어떤 메서드가 다른 곳에서 호출될까 걱정하기보다는 오히려 재사용이 권장하는 코드를 작성하는 것이 좋지 않나 생각합니다. 재사용하기 좋은 코드란 **내부 구현이 변경되더라도 기능을 사용하는 곳의 영향을 최소화 하는 것**이고 이는 곧 유연성있는 코드를 뜻합니다. 유연성있는 코드를 작성하기 위해 사용하는 것이 캡슐화입니다.

캡슐화를 위해선 두 가지 규칙을 사용할 수 있습니다.

- Tell, Don’t Ask
- 데미테르의 법칙

### Tell, Don’t Ask

Tell, Don’t Ask 규칙이란 데이터를 물어보지 않고 기능의 실행만을 요청하는 것입니다.

```java
if(member.getExpiryDate().getDate() < System.currentTimeMillis()) {
	//만료 되었을 때의 처리
}
```
(캡슐화가 되지 않은 코드)


```java
if(member.isExpired()) {
	//만료 되었을 때의 처리
}
```
(캡슐화된 코드)


이렇게 데이터가 아닌 기능을 요구하는 것으로 해당 기능의 구현을 캡슐화 하고 있습니다. 만일 만료 여부를 따지는 로직이 변경되더라도 캡슐화된 기능을 사용하는 코드의 수정은 불필요해질 것입니다.

### 데미테르의 법칙

데미테르의 법칙은 Tell, Don’t Ask규칙을 지킬수 있도록 도와주는 규칙입니다. 데미테르의 법칙은 아래 세가지 규칙으로 이루어집니다.

- 메서드에서 생성한 객체의 메서드만 호출
- 파라미터로 받은 객체의 메서드만 호출
- 필드로 참조하는 객체의 메서드만 호출

```java
if(member.getDate().getTime() < ...) { // 법칙 위반
```
(캡슐화가 되지 않은 코드)


```java
if(member.getTimeFromDate() < ...) { 
```
(캡슐화 된 코드)

메서드의 비유연성은 해당 메서드가 다른 서비스에서 온 것이기 때문에 발생하는 것이 아니라고 생각합니다.

## 데이터 은닉에 관한 문제

다른 service를 참조한다고 해서 모든 service의 메서드를 public으로 만들 필요는 없다 생각합니다. 일부 메서드가 public하게 되는 건 다른 도메인과 자주 상호작용이 일어나는 권한, 팀원, 텃밭 정도가 되지 않을까요??

그중에서도 일부는 Controller에서 사용되기에 이미 public하게 사용될 메서드라 생각됩니다.

또한 제가 제시한 방법은 신뢰성있는 객체를 위한 것이기도 하기에 메서드에서 순수한 객체가 아니라 DTO를 반환하길 권장하는 바입니다. 따라서 데이터 은닉에 대해서는 크게 걱정 안해도 될 것 같다 생각합니다. DTO를 통해 객체값이 변할 일은 없을테니까요.

## 순환 참조 문제

전에 미나님이 작성해주신 리뷰에 제가 첨부한 포스팅에도 동일한 문제에 대해 언급하고 있습니다.

[Service에서 다른 Service 의존에 대하여](https://jangjjolkit.tistory.com/62)

위 포스팅에서도 저의와 동일한 문제에 대해 고민하고 있는데 

1. Service는 무조건 DAO와 의존관계 가지기
2. Controller에서 여러 Service 기능을 호출하기
3. Service layer를 구분하여 Service 간 의존하기

위 세가지 사항에 대해 고려하고 있습니다.

 이때 3번 방식은 서비스 레이어의 의존성 방향을 정하고 파사드 패턴 (Facade Pattern)을 사용하는 것으로 순환 참조를 방지합니다.

그러나 사실 그냥 생각없이 서비스 레이어를 참조하더라도 저희가 도메인 설계를 잘 했다면 순환참조 문제는 발생하지 않지 않을까 생각합니다... 저희가 의도적으로 나눠둔 것은 아니지만 Member나 Role과 같은 Service는 명백하게 Study나 Team과 같은 도메인에는 의존할 일이 없으니까요. 

Curriculum같은 부분을 걱정하실수도 있을 거 같은데 이 부분은 저희가 Study라는 하나의 큰 도메인으로 묶어놓은 상태라 어차피 Service 참조가 아닌 DAO 참조를 할수밖에 없을 겁니다. (Study 객체를 생성하기 위해선 CurriculumItem객체가 필요함)

## 결론

언급했다 싶이 제가 하고 싶은 이야기는 **의존성 & 결합도에 대해** 부분에 다 적혀있습니다. 

일단 요약해보자면 아래와 같습니다.

- Service의 의존이 없으면 중복되는 코드가 많아지고 그만큼 관리 포인트가 많아진다.
- Repository에서의 직접 조회는 순수한 객체를 반환하며 객체의 신뢰성이 낮아진다.
- 서로 다른 도메인에서는 Service를 참조하는 것을 기본으로 하되 Study,Curriculum. Memberm,Role과 같이 동일한 도메인에 속한 경우 필요에 따라 Repository 참조를 가능하도록 한다(할 수밖에 없다)

**Comment by JJimini** [comment]
comment 잘 읽어보았습니다!😊
제가 파악한 팜님의 의견은 다음과 같습니다!
- 본문의 의존성이 낮아진다는 것은 불필요한 객체 할당이 사라지며 의존성이 줄어든다는 의미이다.
- 제가 제시한 방법은 의존 대상이 Service에서 Repository로 바뀌었을 뿐 의존이 발생하는 것은 동일하다.

## 의존성에 대해
팜님의 말씀처럼 '불필요한 객체 할당'으로 인한 의존성 감소는 이해하였습니다! 하지만 제가 앞서 말씀드렸던 **컴파일타임 의존성은 여전히 해결되지 않은 것**으로 보입니다. 예를 들어 **`Team Service`를 의존하는 모든 `Service`를 컴파일 타임 때 다 알아야 하며, 그 의존된 코드가 의존하고 있는 `Service`를 또 다 알아야 하는 연쇄 의존 작용**이 생긴다는 의미입니다!!

## 의존이 발생하는 것은 동일하다.
객체 지향을 정의하면, **의존성 관리**입니다.
따라서 **의존은 발생할 수 밖에 없으며 그 의존을 어떻게 잘 관리하느냐가 중점**이 된다는 의미입니다. 저희는 의존성을 잘 관리할 수 있는 방법을 찾아보아야 할 것 같습니다!! (프로젝트에 최대한 단점이 생기지 않는 방향으로!)
- 중복되는 코드는 그 코드를 처리하는 다른 service class를 만들어서 그 service를 의존하는 방법
    - service의 역할과 책임이 줄어든다.
    - 상대적으로 가벼운 service를 의존할 수 있다. 
- 객체의 신뢰성 문제는 엔티티에 해당 내용을 따로 빼서 해결할 수도 있을 것 같은데요...?!
등등

## 테스트의 어려움
저희도 있었던 상황인 것 같은데요, curriculumItem에서 cropCommandService를 호출하며 crop 테스트 작성 시 curriculumItem에서 진행 + curriculumItem 상황이 다 갖추어 져 있어야 하는 문제가 발생한 적이 있었습니다! 😅😅
Controller와 연결되어 있는 Service 코드를 재사용한다면, 테스트 해야할 위치도 애매하고 테스트코드의 길이가 길어져 테스트 내용을 확인하기 어렵다는 단점이 있을 것 같습니다..!

## 참조해주신 링크들
우선 첫번째 링크는 제가 말씀드렸던 File이라는 특이 케이스인 것 같습니다.
두번째 링크는 제가 제안했던 방식의 convenience 형태의 참조인 것 같습니다.
또한 두번째 링크에 있는 MemberService에 있는 코드는 Repository를 조회하는 코드로, 이름만 Service일뿐 convenience에 들어갈 코드와 동일하다고 생각하며 저희와 상황이 다르다고 생각합니다!

## 결론
- **객체 할당 의존성 문제는 더 큰 의미의 컴파일 의존성과 결합도가 커지는 문제가 여전히 발생하고 있다는 점**
- **Service간 의존성이 생기는 건 그 의존성을 잘 관리해야 하는 것이 개발자의 역량**

다음과 같은 이유로 여전히 **같은 일을 처리하는 Service(각 Controller에서 사용되는 서비스 메서드)를 의존하는 방향으로 발생하는 문제를 해결해야 하는가** 에 대해 좀 더 생각해보아야 할 것 같다는 의견입니다!


저도 여러가지 책이나 링크를 참조하고 싶었지만, 시간 이슈로 인해 제가 백엔드를 공부하며 정리해두었던 자료를 참고로 말씀드린 점 양해부탁드립니다.😢😢😢

(저도 여기까지가 제 답변이며 이후 내용은 팜님이 추가로 작성하셨던 부분에 대한 답변입니다.)

---

## 특정 서비스의 메서드에서 발생한 문제가 해당 메서드를 사용한 다른 서비스에 영향을 미치는 경우에 대해

안정성 vs 관리포인트에 관한 말씀이신 것 같습니다. 이 부분은 저희가 간접 참조 방식을 사용하고 있는 부분을 예시로 들 수 있을 것 같아요~! 저희가 간접 참조를 이용함으로 인해 직접 삭제 로직을 다 구현해야하는 일이 발생하였습니다 _(관리포인트가 많아지는 일)_ . 하지만 저희는 **데이터베이스 최적화를 위해 관리포인트가 늘어나도 해당 방법을 채택**하였습니다! 저는 그게 저희 서비스의 핵심이라고 생각합니다. 이 부분은 human issue가 발생할 위험이 크지만, 프로젝트 단위로써는 이득이 큰 방법을 선택하였던 것 처럼요!

## 여전한 중복의 발생
`createStudy`의 검증 로직을 삭제하면 검증 로직까지 가기 전까지 **불필요하게 실행되는 코드와 DB 조회**와 생기기 때문에 문제가 생길 수 있을 것 같아요!

## 메서드 유연성에 관해
이 부분에서 말씀하신 2가지 원칙은 저도 처음 들어보는 거라 공부해보겠습니다!😅😅

## 데이터 은닉 관한 문제
저도 이 부분에 대해서는 이미 `private`으로 선언되어있는 부분을 사용 해야 하는 경우를 특정하여 말씀드린 부분이었습니다! 하지만 **저희는 이미 DTO(Record)를 통해 데이터를 받아오고 final로 불변을 명시하고 있는데, 불변을 위해 DTO를 반환 해야 하는 필요성은 잘 모르겠습니다😢** 또한 DTO는 Data를 Transfer하는 역할로, 제시해주신 코드에서는 "_클라이언트와 서버 간  Data Transfer_" 의미로 사용되지 않는 것 같습니다! DTO를 반환해야하는 다른 이유가 있다면 알려주세요!

## 순환 참조 문제
제가 잘 이해한 건지는 모르겠으나 해당 링크의 3번에서 말한 방식이 제가 제안드린 방식 같은 느낌이었습니당!
